docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,616 @@
+the domainDB is responsible for storing it.
+
+A domainDB is the lowest level of abstraction: it provides means to write and read data from
+the file system, and maintains the index (see concepts).
+
+Each layer has their own entities and methods of reads, writes, deletes, and garbage collection;
+all interactions are passed down to the domain level to interact with the index and file system.
+
+## 4.1 Top-down view of race conditions in Cesium
+
+Protecting the integrity of data in Cesium is crucial.
+The three shared resources that we cannot allow concurrent access to are data stored on the file system,
+the index stored in memory, and its persisted counterpart on the file system.
+We will spend most of our time talking about the first two,
+since the persisted index is never read from until the database is closed and reopened.
+
+Each one of the read, write, delete, and garbage collection operations is a combination
+of accesses to the index and telemetry data. In the following diagram, we consider the
+life cycle of each of these operations and which resources they need to access.
+
+<div align="center">
+    <img src="img/0019-240529-cesium-race-conditions/Operation-Entities.png" width="90%" />
+    <h6>Cesium entities and resources</h6>
+</div>
+
+We can see that every operation uses the telemetry data in the file system,
+the index, or one of the two – additionally, if the index is modified,
+it must be persisted to the file system as well.
+Throughout this RFC, we will discuss which red-coloured operations can occur at the same time,
+and how, and which ones cannot.
+The general principle we will abide by is that while one goroutine modifies some information,
+no one else should read or write from it.
+If no goroutines are writing to the resource but all are merely reading from it,
+that is acceptable, since all goroutines read the same information.
+
+### 4.1.1 Contentions with reading
+
+Read operations are carried out through an iterator, which only reads one chunk of telemetry
+data at a time.
+This means that every time the caller reads a block of telemetry data,
+it first gets the information about the location and length of the data in the file system,
+then uses that information to read on disk.
+
+With this approach, the caller must understand that if the index was changed between two domains –
+say, the iterator was on position 1, and on a call to `Next()`, it moves on to position 2.
+However, a writer inserts a new domain between the two, meaning that the iterator thinks it is on position 2,
+when in reality it loaded the domain at position 3.
+
+Other than this race condition with the index that the user must beware of,
+read does not conflict with any other operations as it does not change the index nor the underlying data.
+
+### 4.1.2 Write-Delete contention
+
+Deletion works by rewriting the pointers where the specified time range's start and end
```

Comment:
It could be better to explain directly underneath the pictures above how reading, writing, deletion, and garbage collection works, and then in these only cover all possible contentions between each of the four operations.